src/main/java/com/internet/mixshop/dao/jdbc/ShoppingCartDaoJdbcImpl.java

```diff
@@ -0,0 +1,178 @@
+import com.internet.mixshop.lib.Dao;
+import com.internet.mixshop.model.Product;
+import com.internet.mixshop.model.ShoppingCart;
+import com.internet.mixshop.util.ConnectionUtil;
+import java.sql.Connection;
+import java.sql.PreparedStatement;
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import java.sql.Statement;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+
+@Dao
+public class ShoppingCartDaoJdbcImpl implements ShoppingCartDao {
+    @Override
+    public Optional<ShoppingCart> getByUserId(Long userId) {
+        ShoppingCart shoppingCart = null;
+        try (Connection connection = ConnectionUtil.getConnection()) {
+            String query = "SELECT * FROM shopping_carts "
+                    + "WHERE user_id = ? AND is_deleted = false;";
+            PreparedStatement statement = connection.prepareStatement(query);
+            statement.setLong(1, userId);
+            ResultSet resultSet = statement.executeQuery();
+            if (resultSet.next()) {
+                Long cartId = resultSet.getLong("cart_id");
+                shoppingCart = new ShoppingCart(userId);
+                shoppingCart.setId(cartId);
+                statement.close();
+            }
+        } catch (SQLException e) {
+            throw new DataProcessingException("Can't find the shopping cart for user id "
+                    + userId, e);
+        }
+        shoppingCart.setProducts(getProductsInCart(shoppingCart.getId()));
+        return Optional.ofNullable(shoppingCart);
+    }
+
+    @Override
+    public ShoppingCart create(ShoppingCart cart) {
+        try (Connection connection = ConnectionUtil.getConnection()) {
+            String query = "INSERT INTO shopping_carts (user_id) VALUES (?);";
+            PreparedStatement statement =
+                    connection.prepareStatement(query, Statement.RETURN_GENERATED_KEYS);
+            statement.setLong(1, cart.getUserId());
+            statement.executeUpdate();
+            ResultSet resultSet = statement.getGeneratedKeys();
+            if (resultSet.next()) {
+                cart.setId(resultSet.getLong(1));
+            }
+            statement.close();
+            addProductsToCart(cart, connection);
+            return cart;
+        } catch (SQLException e) {
+            throw new DataProcessingException("Can't create " + cart, e);
+        }
+    }
+
+    @Override
+    public Optional<ShoppingCart> getById(Long id) {
+        try (Connection connection = ConnectionUtil.getConnection()) {
+            String query = "SELECT * FROM shopping_carts "
+                    + "WHERE user_id = ? AND is_deleted = false;";
+            PreparedStatement statement = connection.prepareStatement(query);
+            statement.setLong(1, id);
+            ResultSet resultSet = statement.executeQuery();
+            if (resultSet.next()) {
+                return Optional.of(getCartFromResult(resultSet, connection));
```

Comment:
here you call `getCartFromResult()` inside the open connection
`getCartFromResult()` calls `getProductsInCart()` where another connection is opened

we have two connections one inside of another - let's try to avoid it